packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -5076,21 +5076,79 @@ class _ScribblePlaceholder extends WidgetSpan {
   }
 }
 
-/// A text boundary that uses code units as logical boundaries.
+/// A text boundary that uses code points as logical boundaries.
 ///
-/// This text boundary treats every character in input string as an utf-16 code
-/// unit. This can be useful when handling text without any grapheme cluster,
-/// e.g. password input in [EditableText]. If you are handling text that may
-/// include grapheme clusters, consider using [CharacterBoundary].
-class _CodeUnitBoundary extends TextBoundary {
-  const _CodeUnitBoundary(this._text);
+/// A code point represents a single character. This may be smaller than what is
+/// represented by a user-perceived character, or grapheme. For example, a
+/// single grapheme (in this case a Unicode extended grapheme cluster) like
+/// "👨‍👩‍👦" consists of five code points: the man emoji, a zero
+/// width joiner, the woman emoji, another zero width joiner, and the boy emoji.
+/// The [String] has a length of eight because each emoji consists of two code
+/// units.
+///
+/// Code units are the units by which Dart's String class is measured, which is
+/// encoded in UTF-16.
+///
+/// See also:
+///
+///  * [String.runes], which deals with code points like this class.
+///  * [String.characters], which deals with graphemes.
+///  * [CharacterBoundary], which is a [TextBoundary] like this class, but whose
+///    boundaries are graphemes instead of code points.
+class _CodePointBoundary extends TextBoundary {
+  const _CodePointBoundary(this._text);
 
   final String _text;
 
+  // Returns true if the given position falls in the middle of a single code
+  // point in _text.
+  bool _breaksCodePoint(int position) {
```

Comment:
Looks like this function tests whether `position` points to a low surrogate? If that's the case maybe name it as such. Also it's probably easier to just `return _text[position] & 0xFC00 == 0xDC00`.

Reply:
It also requires that the following code unit is a high surrogate.  So in other words, it returns true if `position` is in the middle of a surrogate pair.

I'm changing it to this, let me know what you think:

```dart
  // Returns true if the given position falls in the center of a surrogate pair.
  bool _breaksSurrogatePair(int position) {
    assert(position > 0 && position < _text.length && _text.length > 1);
    return _text.codeUnitAt(position) & 0xFC00 == 0xDC00
        && _text.codeUnitAt(position + 1) & 0xFC00 == 0xD800;
  }
```

Reply:
In UTF16 encoding a surrogate pair has exactly 2 code units, a high surrogate followed by a low surrogate. Otherwise it's an encoding error. I think you meant to check if `position` points to a low surrogate?

Reply:
Whoops yes, I had that backwards and I was using `position` incorrectly.  I've fixed it and the tests seem to pass now.

I also reused this same low/high surrogate logic from TextPainter by making two methods public.